winusb.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2016, Devan Lai
+ *
+ * Permission to use, copy, modify, and/or distribute this software
+ * for any purpose with or without fee is hereby granted, provided
+ * that the above copyright notice and this permission notice
+ * appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+ * LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <libopencm3/usb/usbd.h>
+#include "winusb.h"
+
+static struct winusb_compatible_id_descriptor winusb_wcid = {
+	.header = {
+		.dwLength = sizeof(struct winusb_compatible_id_descriptor_header) +
+					1 * sizeof(struct winusb_compatible_id_function_section),
+		.bcdVersion = WINUSB_BCD_VERSION,
+		.wIndex = WINUSB_REQ_GET_COMPATIBLE_ID_FEATURE_DESCRIPTOR,
+		.bNumSections = 1,
+		.reserved = { 0, 0, 0, 0, 0, 0, 0 },
+	},
+	.functions = {
+		{
+			.bInterfaceNumber = 2,
+			.reserved0 = { 1 },
+			.compatibleId = "WINUSB",
+			.subCompatibleId = "",
+			.reserved1 = { 0, 0, 0, 0, 0, 0}
+		},
+	}
+};
+
+static struct usb_string_descriptor winusb_string_descriptor = {
+  .bLength = sizeof(WINUSB_EXTRA_STRING) + sizeof(struct usb_string_descriptor),
+  .bDescriptorType = USB_DT_STRING,
+  .wData = WINUSB_EXTRA_STRING
+};
+
+static const struct winusb_extended_properties_descriptor guid = {
+	.header = {
+		.dwLength = sizeof(struct winusb_extended_properties_descriptor_header)
+					+ 1 * sizeof (struct winusb_extended_properties_feature_descriptor),
+		.bcdVersion = WINUSB_BCD_VERSION,
+		.wIndex = WINUSB_REQ_GET_EXTENDED_PROPERTIES_OS_FEATURE_DESCRIPTOR,
+		.wNumFeatures = 1,
+	},
+	.features = {
+		{
+			.dwLength = sizeof(struct winusb_extended_properties_feature_descriptor),
+			.dwPropertyDataType = WINUSB_EXTENDED_PROPERTIES_MULTISZ_DATA_TYPE,
+			.wNameLength = sizeof(u"DeviceInterfaceGUIDs"),
+			.name = u"DeviceInterfaceGUIDs",
+			.dwPropertyDataLength = (WINUSB_EXTENDED_PROPERTIES_GUIDS_SIZE_C),
+			.propertyData = WINUSB_EXTENDED_PROPERTIES_RANDOM_GUID,
+		},
+	}
+};
+
+static int winusb_descriptor_request(usbd_device *usbd_dev,
+					struct usb_setup_data *req,
+					uint8_t **buf, uint16_t *len,
+					usbd_control_complete_callback* complete) {
+	(void)complete;
+	(void)usbd_dev;
+
+	if ((req->bmRequestType & USB_REQ_TYPE_TYPE) != USB_REQ_TYPE_STANDARD) {
+		return USBD_REQ_NEXT_CALLBACK;
+	}
+
+	if (req->bRequest == USB_REQ_GET_DESCRIPTOR && req->wValue >> 8 == USB_DT_STRING) {
```

Comment:
Use `usb_descriptor_type(req->wValue)`

Reply:
those are not exported from usb_standard.c; I didn't want to retype them here for 1 use :)

Reply:
but maybe yeah, it will make it more readable

Reply:
Oops, my bad. I was looking at http://libopencm3.org/docs/latest/stm32f2/html/group__usb__standard__file.html and I didn't see it said `static int`.

Reply:
I have added those and used them. It doesn't hurt. (Maybe macro would be quicker, but maybe it won't matter.)